pinot-controller/src/main/java/org/apache/pinot/controller/recommender/rules/impl/SegmentSizeRule.java

```diff
@@ -0,0 +1,142 @@
+import org.apache.pinot.controller.recommender.rules.AbstractRule;
+import org.apache.pinot.controller.recommender.rules.io.configs.SegmentSizeRecommendations;
+import org.apache.pinot.controller.recommender.rules.io.params.SegmentSizeRuleParams;
+import org.apache.pinot.spi.config.table.TableConfig;
+import org.apache.pinot.spi.config.table.TableType;
+import org.apache.pinot.spi.data.Schema;
+import org.apache.pinot.spi.utils.builder.TableConfigBuilder;
+
+import static org.apache.pinot.controller.recommender.rules.io.params.RecommenderConstants.SegmentSizeRule.NOT_PROVIDED;
+
+
+/**
+ * This rule generates a segment based on the provided schema characteristics and then recommends the followings
+ * using the size and number of records in the generated segments:
+ *   - number of segments
+ *   - number of records in each segment
+ *   - size of each segment
+ */
+public class SegmentSizeRule extends AbstractRule {
+
+  static final int MEGA_BYTE = 1024 * 1024;
+
+  public SegmentSizeRule(InputManager input, ConfigManager output) {
+    super(input, output);
+  }
+
+  @Override
+  public void run()
+      throws InvalidInputException {
+
+    if (_input.getTableType().equalsIgnoreCase("REALTIME")) {
+      // no need to estimate segment size & optimal number of segments for realtime only tables;
+      // RT Provisioning Rule will have a comprehensive analysis on that
+      return;
+    }
+
+    long segmentSize;
+    int numRows;
+    SegmentSizeRuleParams segmentSizeRuleParams = _input.getSegmentSizeRuleParams();
+    if (segmentSizeRuleParams.getActualSegmentSizeMB() == NOT_PROVIDED
+        && segmentSizeRuleParams.getNumRowsInActualSegment() == NOT_PROVIDED) {
+
+      // generate a segment
+      TableConfig tableConfig = createTableConfig(_input.getSchema());
+      int numRowsInGeneratedSegment = segmentSizeRuleParams.getNumRowsInGeneratedSegment();
+      File generatedSegmentDir =
+          new MemoryEstimator.SegmentGenerator(_input._schemaWithMetaData, _input._schema, tableConfig,
+              numRowsInGeneratedSegment, true).generate();
+      segmentSize = FileUtils.sizeOfDirectory(generatedSegmentDir);
+      numRows = numRowsInGeneratedSegment;
+
+      // cleanup
+      try {
+        FileUtils.deleteDirectory(generatedSegmentDir);
+      } catch (IOException e) {
+        throw new RuntimeException("Cannot delete the generated segment directory", e);
+      }
+    } else {
+      segmentSize = segmentSizeRuleParams.getActualSegmentSizeMB() * MEGA_BYTE;
+      numRows = segmentSizeRuleParams.getNumRowsInActualSegment();
+    }
+
+    // estimate optimal segment count & size parameters
+    SegmentSizeRecommendations params =
+        estimate(segmentSize, segmentSizeRuleParams.getDesiredSegmentSizeMb() * MEGA_BYTE, numRows,
+            _input.getNumRecordsPerPush());
+
+    // wire the recommendations
+    _output.setSegmentSizeRecommendations(params);
+    _input.capCardinalities((int) params.getNumRows());
```

Comment:
(nit) suggest adding a comment explaining that computed CPS will then be updated in the InputManager to be used by subsequent rules after this one ....

Reply:
Done.